File: kata-8-kyu/add-length/python/add_length_test.py
# + =====================================================================================================================
# +
# + 8 kyu - Add Length  [ ID: 559d2284b5bb6799e9000047 ] (add-length)
# + URL: https://www.codewars.com/kata/559d2284b5bb6799e9000047
# + Category: REFERENCE  |  Tags: ARRAYS | LISTS | FUNDAMENTALS
# +
# + =====================================================================================================================

import codewars_test as test
from add_length import add_length


@test.describe("Fixed Tests")
def basic_tests():
    test.assert_equals(add_length("apple ban"), ["apple 5", "ban 3"])
    test.assert_equals(add_length("you will win"), ["you 3", "will 4", "win 3"])
    test.assert_equals(add_length("you"), ["you 3"])
    test.assert_equals(add_length("y"), ["y 1"])
    test.assert_equals(add_length("x y z"), ["x 1", "y 1", "z 1"])
    test.assert_equals(add_length("xyz"), ["xyz 3"])
    test.assert_equals(add_length("xyz x y z xyz"), ["xyz 3", "x 1", "y 1", "z 1", "xyz 3"])
    test.assert_equals(add_length("a bc cde"), ["a 1", "bc 2", "cde 3"])
    test.assert_equals(add_length("a ab abc"), ["a 1", "ab 2", "abc 3"])
    test.assert_equals(add_length("a ab abc ab a"), ["a 1", "ab 2", "abc 3", "ab 2", "a 1"])


@test.describe("Random Tests")
def random_tests():

    from random import randint
    from re import sub

    add_sol = lambda s: [*map(lambda x: " ".join([x, str(len(x))]), s.split(" "))]

    for _ in range(40):
        base = "    abcdefghijklmnopqrstuvwxyz    ABCDEFGHIJKLMNOPQRSTUVWXYZ    "
        s = "".join([base[randint(0, len(base) - 1)] for _ in range(randint(1, 60))]).strip()
        s = sub("\s{2,100}", " ", s)
        if s == "":
            s = "CodeWars"

        @test.it(f"Testing for add_length({s})")
        def test_case():
            test.assert_equals(add_length(s), add_sol(s))

File: kata-8-kyu/add-length/javascript/addLength.Test.js
//+ ====================================================================================================================
//+
//+ 8 kyu - Add Length  [ ID: 559d2284b5bb6799e9000047 ] (add-length)
//+ URL: https://www.codewars.com/kata/559d2284b5bb6799e9000047
//+ Category: REFERENCE  |  Tags: ARRAYS | LISTS | FUNDAMENTALS
//+
//+ ====================================================================================================================

import { assert } from "vitest"
const { addLength } = require("./addLength")

it("Basic tests", function () {
  assert.deepEqual(addLength("apple ban"), ["apple 5", "ban 3"])
  assert.deepEqual(addLength("you will win"), ["you 3", "will 4", "win 3"])
  assert.deepEqual(addLength("you"), ["you 3"])
  assert.deepEqual(addLength("y"), ["y 1"])
  assert.deepEqual(addLength("x y z"), ["x 1", "y 1", "z 1"])
  assert.deepEqual(addLength("xyz"), ["xyz 3"])
  assert.deepEqual(addLength("xyz x y z xyz"), ["xyz 3", "x 1", "y 1", "z 1", "xyz 3"])
  assert.deepEqual(addLength("a bc cde"), ["a 1", "bc 2", "cde 3"])
  assert.deepEqual(addLength("a ab abc"), ["a 1", "ab 2", "abc 3"])
  assert.deepEqual(addLength("a ab abc ab a"), ["a 1", "ab 2", "abc 3", "ab 2", "a 1"])
})

describe("Random tests", function () {
  function randint(a, b) {
    return Math.floor(Math.random() * (b - a + 1) + a)
  }
  const base = "    abcdefghijklmnopqrstuvwxyz    ABCDEFGHIJKLMNOPQRSTUVWXYZ    "
  function add_sol(str) {
    return str.split(" ").map(function (a) {
      return [a, a.length].join(" ")
    })
  }

  for (let _ = 0; _ < 40; _++) {
    let s = [],
      len = randint(1, 60)
    for (let j = 0; j < len; j++) {
      s.push(base[randint(0, base.length - 1)])
    }
    s = s
      .join("")
      .trim()
      .replace(/\s{2,100}/g, " ")
    if (s == "") s = "CodeWars"
    it("Testing for " + JSON.stringify(s), function () {
      assert.deepEqual(addLength(s), add_sol(s), "It should work for random inputs too")
    })
  }
})
